### PR TITLE
PBF

### PR DESCRIPTION
**Pull request by WhisperWind22**
Plane of Best Fit (https://github.com/rdkit/rdkit/tree/master/Contrib/PBF, https://pubs.acs.org/doi/abs/10.1021/ci300293f)

### CODE REVIEW

File: mordred/PBF.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    order = np.argsort(vals)
+    normal = vects[:, order[0]]
+    plane = np.zeros((4,), np.double)
+    plane[:3] = normal
+    plane[3] = -1 * normal.dot(origin)
+    return plane
+
+
+def calculate_PBFRD(mol, confId=-1):
+    conf = mol.GetConformer(confId)
```
Comment (philopon) [inline mordred/PBF.py]: Please use `self.coord` in `calculate` method of `Descriptor` instead of GetConformer method.

File: mordred/PBF.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+from numpy import linalg
+
+from ._base import Descriptor
+
+__all__ = (
+    "PBF",
+)
+
+
+def GetBestFitPlane(pts, weights=None):
```
Comment (philopon) [inline mordred/PBF.py]: Use snake_case and `_` prefix, please.

File: mordred/PBF.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        res += np.abs(pt.dot(plane[:3]) + plane[3])
+    res /= denom
+    res /= len(pts)
+    return res
+
+
+class PBF(Descriptor):
+    r"""PBF descriptor.
+    """
+
```
Comment (philopon) [inline mordred/PBF.py]: Please add `require_3D = True` attribute to handle 3D coordinate.

Comment (philopon) [inline mordred/PBF.py]: Please add `__slots__ = ()` attribute.

File: mordred/PBF.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    def preset(cls):
+        yield cls()
+
+    def description(self):
+        return self.__class__.__name__
+
+    def __str__(self):
+        return self.__class__.__name__
+
+    def __init__(self, type="PBF"):
```
Comment (philopon) [inline mordred/PBF.py]: __init__ is needless.

File: mordred/PBF.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    vals, vects = linalg.eigh(sums)
+    order = np.argsort(vals)
+    normal = vects[:, order[0]]
+    plane = np.zeros((4,), np.double)
+    plane[:3] = normal
+    plane[3] = -1 * normal.dot(origin)
+    return plane
+
+
+def calculate_PBFRD(mol, confId=-1):
```
Comment (philopon) [inline mordred/PBF.py]: Add `_` prefix to internal functions, please.

File: mordred/tests/test_PBF.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
+# Based on the code from https://github.com/rdkit/rdkit/blob/master/Contrib/PBF/pbf.py
+import numpy as np
+from numpy import linalg
+from rdkit import Chem
+
+
+def test_PBF():
+    def GetBestFitPlane(pts, weights=None):
```
Comment (philopon) [inline mordred/tests/test_PBF.py]: Please import from `mordred/PBF.py`.

File: mordred/tests/test_PBF.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        sums /= wSum
+        vals, vects = linalg.eigh(sums)
+        order = np.argsort(vals)
+        normal = vects[:, order[0]]
+        plane = np.zeros((4,), np.double)
+        plane[:3] = normal
+        plane[3] = -1 * normal.dot(origin)
+        return plane
+
+    def PBFRD(mol, confId=-1):
```
Comment (philopon) [inline mordred/tests/test_PBF.py]: Please import from mordred/PBF.py.

File: mordred/tests/test_PBF.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        res = 0.0
+        for pt in pts:
+            res += np.abs(pt.dot(plane[:3]) + plane[3])
+        res /= denom
+        res /= len(pts)
+        return res
+
+    suppl = Chem.SDMolSupplier('./testData/egfr.sdf', removeHs=False)
+    expected = open('./testData/egfr.out', 'r')
+    for m in suppl:
```
Comment (philopon) [inline mordred/tests/test_PBF.py]: Please:

* create reference dict and access by name.
* move `testData` to `references`
* use __file__ to access data file.
* use Descriptor instance for integrated test (of cause, unit test is also welcome).

example:

```.py
data_dir = os.path.join(
    os.path.dirname(__file__),
    "references",
)

references = {}
for line in open(os.path.join(data_dir, "egfr.out")):
    name, value = line.strip().split()
    references[name] = float(value)

desc = PBF()

for mol in Chem.SDMolSupplier(os.path.join("egfr.sdf"), removeHs=False):
    result = desc(mol)
    expected = references[mol.GetProp("_Name")]
    yield assert_almost_equal, actual, expected
```

File: mordred/PBF.py (hunk 3 above)
Comment (philopon) [reply mordred/PBF.py]: and add `since = "1.1.0"` attribute, please (it is used preset versionning in current develop branch).

**Comment by philopon** [comment]
Why don't you wrap [rdMolDescriptors.CalcPBF](http://rdkit.org/Python_Docs/rdkit.Chem.rdMolDescriptors-module.html#CalcPBF)?

**Comment by WhisperWind22** [comment]
Sorry for the long answer. I just can not make Travis work.

**Comment by philopon** [comment]
> Sorry for the long answer.
Not at all 😄 

> I just can not make Travis work.
> Why do you use the private key in the Travis environment variables?
The private key is used in upload to anaconda cloud.
And it sould be fixed in #36 (it already merged in latest develop branch).
 I'm sorry for your inconvenience but could you please rebase to latest develop?

Thanks,

**Comment by WhisperWind22** [comment]
I made a rebase, but Travis still does not work.
Throws error "iv undefined"

**Comment by philopon** [comment]
Thank you.

It seems that tests are executed and some tests are failed.

1. Please use __file__ attribute instead of relative path.
2. Please add `since` and `__slots__` attribute (it is required in develop branch).
3. Please update `mordred/descriptors/__init__.py` doccomment

Please see review comments for details.

**Comment by WhisperWind22** [comment]
Use the descriptor from rdkit

**Comment by philopon** [comment]
I merged your PR after few modification (d65d3fa451aca3f32adf4124a83532978ae57e46).

thanks!

**Pull request merged**